### PR TITLE
Bug 1905100: Add "ingress.operator.openshift.io/hard-stop-after" annotation

### PR DESCRIPTION
**Pull request by frobware**
Annotating either an ingresscontroller or the ingress config with this
new annoation will redeploy the router and that will configure haproxy
to emit the haproxy "hard-stop-after" global option.

An ingresscontroller with a valid annotation set will override
ingresses.config/cluster (if set).

Examples:

Annotating the ingress config:

    $ oc annotate ingresses.config/cluster ingress.operator.openshift.io/hard-stop-after=1h

Annotating the "default" ingresscontroller:

    $ oc -n openshift-ingress-operator annotate ingresscontrollers/default ingress.operator.openshift.io/hard-stop-after=30m

### CODE REVIEW

**Comment by frobware** [comment]
/hold

Will push some e2e tests.

File: pkg/operator/controller/ingress/deployment.go [hunk 1]
```diff
@@ -136,6 +139,28 @@ func HTTP2IsEnabled(ic *operatorv1.IngressController, ingressConfig *configv1.In
 	return configHasHTTP2Enabled
 }
 
+// HardStopAfterIsEnabledByAnnotation returns true if the map m has
+// the key RouterHardStopAfterEnvName and its value is not the empty
+// string ("").
+func HardStopAfterIsEnabledByAnnotation(m map[string]string) (bool, string) {
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:145]: No need to export this, is there?  Same for `HardStopAfterIsEnabled`.  (But `HTTP2IsEnabledByAnnotation` is already public, and anyway it's a minor thing, so I won't object if you prefer to keep this as-is.)
```suggestion
// hardStopAfterIsEnabledByAnnotation returns true if the map m has
// the key RouterHardStopAfterEnvName and its value is not the empty
// string ("").
func hardStopAfterIsEnabledByAnnotation(m map[string]string) (bool, string) {
```

File: pkg/operator/controller/ingress/deployment.go [hunk 2]
```diff
@@ -136,6 +139,28 @@ func HTTP2IsEnabled(ic *operatorv1.IngressController, ingressConfig *configv1.In
...
+	if val, ok := m[RouterHardStopAfterAnnotation]; ok && len(val) > 0 {
+		return true, val
+	}
+	return false, ""
+}
+
+// HardStopAfterIsEnabled returns true if either the ingress
+// controller or the ingress config has the "hard-stop-after"
+// annotation. The presence of the annoation on the ingress
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:154]:
```suggestion
// annotation. The presence of the annotation on the ingress
```

File: pkg/operator/controller/ingress/deployment.go [hunk 3]
```diff
@@ -623,6 +648,10 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
 		env = append(env, corev1.EnvVar{Name: RouterDisableHTTP2EnvName, Value: "true"})
 	}
 
+	if enabled, value := HardStopAfterIsEnabled(ci, ingressConfig); enabled && len(value) > 0 {
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:651]: `HardStopAfterIsEnabled` already returns `false` if `len(value) == 0`.
```suggestion
	if enabled, value := HardStopAfterIsEnabled(ci, ingressConfig); enabled {
```

File: pkg/operator/controller/ingress/deployment_test.go [hunk 4]
```diff
@@ -522,6 +522,7 @@ func TestDesiredRouterDeployment(t *testing.T) {
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_UNIQUE_ID_FORMAT", true, `"foo"`)
 
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_H1_CASE_ADJUST", false, "")
+	checkDeploymentHasEnvVar(t, deployment, RouterHardStopAfterEnvName, false, "")
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:525]: We should add test coverage somewhere in this unit test for when the annotation is set.

**Review by Miciah — COMMENTED** [review]
Just some very minor suggestions Overall, it looks good.

File: pkg/operator/controller/ingress/deployment_test.go (hunk 4 above)
Comment (frobware) [reply pkg/operator/controller/ingress/deployment_test.go:525]: > We should add test coverage somewhere in this unit test for when the annotation is set.

In progress.

File: pkg/operator/controller/ingress/deployment.go (hunk 3 above)
Comment (frobware) [reply pkg/operator/controller/ingress/deployment.go:651]: This was belt-and-braces in lieu of any future change. Emitting `hard-stop-after` (i.e., no value) would mean haproxy fails to parse the config.  I can drop it, just wary of crash looping haproxy.

File: pkg/operator/controller/ingress/deployment.go (hunk 1 above)
Comment (frobware) [reply pkg/operator/controller/ingress/deployment.go:145]: > No need to export this, is there? Same for `HardStopAfterIsEnabled`. (But `HTTP2IsEnabledByAnnotation` is already public, and anyway it's a minor thing, so I won't object if you prefer to keep this as-is.)

Currently used by e2e tests (not pushed).  If I don't need it then I'll change the scope.

File: pkg/operator/controller/ingress/deployment.go (hunk 3 above)
Comment (Miciah) [reply pkg/operator/controller/ingress/deployment.go:651]: The unit test can verify that this code does the right thing if the annotation is set but has an empty value.

Comment (frobware) [reply pkg/operator/controller/ingress/deployment.go]: Dropped.

**Comment by frobware** [comment]
/hold cancel

I pushed some tests. This is ready for a review.

File: test/e2e/hard_stop_after_test.go [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+	if err := waitForIngressConfigHardStopAfterToBeUnset(client, timeout, ingressConfig); err != nil {
+		return err
+	}
+	if err := waitForIngressControllerHardStopAfterToBeUnset(client, timeout, ic); err != nil {
+		return err
+	}
+	return waitForRouterDeploymentHardStopAfterToBeUnset(client, timeout, routerDeployment)
+}
+
+func waitForIngressConfigHardStopAfterToBeSet(client client.Client, timeout time.Duration, c *configv1.Ingress, expectedDuration time.Duration) error {
```
Comment (sgreene570) [inline test/e2e/hard_stop_after_test.go:180]: could `waitForIngressConfigHardStopAfterToBeSet`, `waitForIngressConfigHardStopAfterToBeUnSet`, `waitForIngressControllerHardStopAfterToBeSet`, & `waitForIngressControllerHardStopAfterToBeUnSet` all be collapsed into one function with additional parameters for object (ingressconfig or ingresscontroller) and a set/unset bool?

Comment (frobware) [reply test/e2e/hard_stop_after_test.go:180]: Funnily enough I did that first time; second time I found it easier to read with the discrete functions.

File: test/e2e/hard_stop_after_test.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+		t.Fatalf("failed to clear hard-stop-after on ingresscontroller: %v", err)
+	}
+	if err := WaitForHardStopAfterUnsetInAllComponents(kclient, hardStopAfterRetryTimeout, ic, ingressConfig, routerDeployment); err != nil {
+		t.Fatalf("some component still has hard-stop-after set: %v", err)
+	}
+}
+
+// NOTE: This test will mutate the default ingresscontroller and the
+// "cluster" ingress configuration.
+func TestRouteHardStopAfterEnableOnIngressControllerHasPriorityOverIngressConfig(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go:94]: I think it's important to semantically to test this at runtime. They don't take a huge amount of time to run. The unit test coverage isn't huge - we just test that by default hard-stop-after is not set, which is not particularly extensive.  There's also some precedence here with the the http/2 enablement.

File: test/e2e/hard_stop_after_test.go [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+	if err := WaitForHardStopAfterUnsetInAllComponents(kclient, hardStopAfterRetryTimeout, ic, ingressConfig, routerDeployment); err != nil {
+		t.Fatalf("test assertions failed: %v", err)
+	}
+
+	// First set hard-stop-after on the ingress config
+	if err := hardStopAfterTestIngressConfig(t, ingressConfig, routerDeployment); err != nil {
+		t.Fatalf("test assertions failed: %v", err)
+	}
+
+	// Secondly, oveerride setting by adding hard-stop-after on the ingresscontroller.
```
[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go]: Fixed.

File: pkg/operator/controller/ingress/deployment.go [hunk 8]
```diff
@@ -623,6 +649,12 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
 		env = append(env, corev1.EnvVar{Name: RouterDisableHTTP2EnvName, Value: "true"})
 	}
 
+	if enabled, value := HardStopAfterIsEnabled(ci, ingressConfig); enabled {
+		if _, err := time.ParseDuration(value); err != nil {
+			env = append(env, corev1.EnvVar{Name: RouterHardStopAfterEnvName, Value: value})
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment.go:654]: `time.ParseDuration` can't parse things that specify days (ie `1d`). We don't expect users to be setting values that high right?

Comment (sgreene570) [reply pkg/operator/controller/ingress/deployment.go:654]: Also if we fail to parse the annotation should we log that?

Comment (Miciah) [reply pkg/operator/controller/ingress/deployment.go:654]: I think the logic is inverted here:  We should use `value` only if `time.ParseDuration(value)` returns a nil error value.

Comment (Miciah) [reply pkg/operator/controller/ingress/deployment.go:654]: Perhaps `HardStopAfterIsEnabledByAnnotation` should perform the validation.  What should happen if `ingresses.config/cluster` has a valid annotation and then the user adds an invalid annotation to the ingresscontroller?  I think the first annotation should remain in effect, but with the current logic, the invalid annotation on the ingresscontroller causes the valid annotation on the ingress.config to be ignored.

File: pkg/operator/controller/ingress/deployment.go [hunk 9]
```diff
@@ -1076,3 +1076,45 @@ func deploymentConfigChanged(current, expected *appsv1.Deployment) (bool, *appsv
 	updated.Spec.Replicas = &replicas
 	return true, updated
 }
+
+// clipHAProxyTimeoutValue prevents the HAProxy config file from using
+// timeout values that exceed the maximum value allowed by HAProxy.
+// Return the parameter string instead of an err in the event that a
+// timeout string value is not parsable as a valid time duration.
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:1083]: But it does return the error value too.

File: pkg/operator/controller/ingress/deployment.go (hunk 8 above)
Comment (frobware) [reply pkg/operator/controller/ingress/deployment.go]: Done and added a test; though I will add an additional test that explicitly sets "" as a timeout value - just need to rearrange the test a bit now.

File: pkg/operator/controller/ingress/deployment.go (hunk 9 above)
Comment (frobware) [reply pkg/operator/controller/ingress/deployment.go]: Updated.

File: test/e2e/hard_stop_after_test.go [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+	if err != nil {
+		t.Fatalf("failed to get test objects: %v", err)
+	}
+
+	if err := hardStopAfterTestIngressConfig(t, ingressConfig, routerDeployment); err != nil {
+		t.Fatalf("test assertions failed: %v", err)
+	}
+
+	// Cleanup
+	if err := clearHardStopAfterDurationForIngressConfig(t, kclient, hardStopAfterRetryTimeout, ingressConfig); err != nil {
```
[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go:56]: This doesn't hold true in terms of moving them above any t.Fatal calls. It's only at the places that say `// cleanup` that you could make them into a defer block so I don't know what we would gain. I will remove the comments. If you want these to be rock solid and have no effect on any other test then the better thing to do would be to test against their own controller, config, deployment - but they will take longer to run.

[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go:56]: /retest

[the remark this answers is not in the document]
Comment (sgreene570) [reply test/e2e/hard_stop_after_test.go]: So if any test assertions fail, then we don't need to revert the default ingress controller back to it's original non-annotated state?

File: test/e2e/hard_stop_after_test.go [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+	if err := hardStopAfterTestIngressController(t, kclient, ic, routerDeployment, (500 * time.Minute).String()); err != nil {
+		t.Fatalf("test assertions failed: %v", err)
+	}
+
+	// Remove the controller setting.
+	if err := setHardStopAfterDurationForIngressController(t, kclient, hardStopAfterRetryTimeout, "", ic); err != nil {
+		t.Fatalf("failed to clear hard-stop-after on ingresscontroller: %v", err)
+	}
+
+	// Deployment should reveet back to config value
```
Comment (sgreene570) [inline test/e2e/hard_stop_after_test.go:93]: nit: reveet -> revert

File: test/e2e/hard_stop_after_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sgreene570) [reply test/e2e/hard_stop_after_test.go]: http2 tests seem to follow a similar pattern to this PR. So maybe we should double check any possible defer improvements in a follow up for all e2e tests in general.

**Comment by sgreene570** [comment]
e2e-aws-operator passed
/lgtm in the interest of time.

File: test/e2e/hard_stop_after_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go]: > defer improvements i

It's not clear to me what a defer improvement would look like.  The test has failed. CI will eventually fail, so there's nothing that would aid getting a green run. Mutating the same objects in multiple tests doesn't help things.

Should defer look like: 

```go
	defer func() {
		if err := setHardStopAfterDurationForIngressController(t, kclient, hardStopAfterRetryTimeout, "", ic); err != nil {
			t.Fatalf("failed to clear hard-stop-after on ingresscontroller: %v", err)
		}
		if err := waitForHardStopAfterUnsetInAllComponents(kclient, hardStopAfterRetryTimeout, ic, ingressConfig, routerDeployment); err != nil {
			t.Fatalf("some component still has hard-stop-after set: %v", err)
		}
	}()

	if err := hardStopAfterTestIngressController(t, kclient, ic, routerDeployment, (300 * time.Minute).String()); err != nil {
		t.Fatalf("test assertions failed: %v", err)
	}
```

Or?

```go
	defer func() {
		if err := waitForHardStopAfterUnsetInAllComponents(kclient, hardStopAfterRetryTimeout, ic, ingressConfig, routerDeployment); err != nil {
			t.Fatalf("some component still has hard-stop-after set: %v", err)
		}
	}()

	defer func() {
		if err := setHardStopAfterDurationForIngressController(t, kclient, hardStopAfterRetryTimeout, "", ic); err != nil {
			t.Fatalf("failed to clear hard-stop-after on ingresscontroller: %v", err)
		}
	}()

	if err := hardStopAfterTestIngressController(t, kclient, ic, routerDeployment, (300 * time.Minute).String()); err != nil {
		t.Fatalf("test assertions failed: %v", err)
	}
```

given that defer runs in LIFO.

**Comment by frobware** [comment]
/hold

Need to redo `TestRouteHardStopAfterTestZeroLengthDuration` as it doesn't exercise setting a `""` value, it just clears the annotation  by virtue of calling `setHardStopAfterDurationForIngressController` which is not really the same thing.

File: test/e2e/hard_stop_after_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (frobware) [reply test/e2e/hard_stop_after_test.go]: FTR, TIL that fatal arranges for all defers to be called. Thank you @sgreene570 .

[the remark this answers is not in the document]
Comment (sgreene570) [reply test/e2e/hard_stop_after_test.go]: I think the former (grouping all cleanup code into one defer block) would make the most sense (if we decide we want to add any defers at all).

**Comment by frobware** [comment]
/hold cancel

@sgreene570 I addressed https://github.com/openshift/cluster-ingress-operator/pull/522#issuecomment-758725438. PTAL.

**Comment by sgreene570** [comment]
> /hold cancel
> 
> @sgreene570 I addressed [#522 (comment)](https://github.com/openshift/cluster-ingress-operator/pull/522#issuecomment-758725438). PTAL.

Looks good. 
/lgtm

**Comment by sgreene570** [comment]
`operator_test.go:402: failed waiting deployment openshift-ingress-operator/default to scale to 3: failed to achieve expected replicas, last observed: 2`
/test e2e-aws-operator

**Comment by frobware** [comment]
> @frobware: #522 failed to apply on top of branch "release-4.6":
> 
> ```
> Applying: Add "ingress.operator.openshift.io/hard-stop-after" annotation
> Using index info to reconstruct a base tree...
> M	pkg/operator/controller/ingress/deployment.go
> M	pkg/operator/controller/ingress/deployment_test.go
> Falling back to patching base and 3-way merge...
> Auto-merging pkg/operator/controller/ingress/deployment_test.go
> CONFLICT (content): Merge conflict in pkg/operator/controller/ingress/deployment_test.go
> Auto-merging pkg/operator/controller/ingress/deployment.go
> error: Failed to merge in the changes.
> hint: Use 'git am --show-current-patch=diff' to see the failed patch
> Patch failed at 0001 Add "ingress.operator.openshift.io/hard-stop-after" annotation
> When you have resolved this problem, run "git am --continue".
> If you prefer to skip this patch, run "git am --skip" instead.
> To restore the original branch and stop patching, run "git am --abort".
> ```

Manual backport here: https://github.com/openshift/cluster-ingress-operator/pull/535

**Pull request merged**